typescript/packages/common-charm/src/charm.ts

```diff
@@ -21,76 +21,88 @@ export type Charm = {
   [key: string]: any;
 };
 
+export const charmSchema: JSONSchema = {
+  type: "object",
+  properties: {
+    [NAME]: { type: "string" },
+    [UI]: { type: "object" },
+  },
+} as const;
+
+export const charmListSchema: JSONSchema = {
+  type: "array",
+  items: { ...charmSchema, asCell: true },
+} as const;
+
 export type StorageType = "remote" | "memory" | "local";
 
 export class CharmManager {
-  private charms: DocImpl<DocLink[]>;
+  private charmsDoc: DocImpl<DocLink[]>;
+  private charms: Cell<Cell<Charm>[]>;
 
   constructor(private storage: Storage) {
-    this.charms = getDoc<DocLink[]>([], "charms");
+    this.charmsDoc = getDoc<DocLink[]>([], "charms");
+    this.charms = this.charmsDoc.asCell([], undefined, charmListSchema);
   }
 
   getReplica(): string | undefined {
     return this.storage.getReplica();
   }
 
-  getCharms(): DocImpl<DocLink[]> {
+  getCharms(): Cell<Cell<Charm>[]> {
     // Start syncing if not already syncing. Will trigger a change to the list
     // once loaded.
-    this.storage.syncCell(this.charms);
+    this.storage.syncCell(this.charmsDoc);
     return this.charms;
   }
 
-  async add(newCharms: DocImpl<any>[]) {
-    await this.storage.syncCell(this.charms);
+  async add(newCharms: Cell<Charm>[]) {
+    await this.storage.syncCell(this.charmsDoc);
     await idle();
 
-    const currentCharmsIds = this.charms.get().map(({ cell }) => JSON.stringify(cell.entityId));
-    const charmsToAdd = newCharms.filter(
-      (cell) => !currentCharmsIds.includes(JSON.stringify(cell.entityId)),
-    );
-
-    if (charmsToAdd.length > 0) {
-      console.log("add charms", charmsToAdd);
-      this.charms.send([
-        ...this.charms.get(),
-        ...charmsToAdd.map((cell) => ({ cell, path: [] }) satisfies DocLink),
-      ]);
-    }
+    newCharms.forEach((charm) => {
+      if (!this.charms.get().some((otherCharm) => otherCharm.equals(charm)))
+        this.charms.push(charm);
+    });
+
+    await idle();
   }
 
-  async get(id: string): Promise<DocImpl<any> | undefined> {
-    await this.storage.syncCell(this.charms);
+  async get(id: string): Promise<Cell<Charm> | undefined> {
+    await this.storage.syncCell(this.charmsDoc);
+
+    const idAsDocId = JSON.stringify({ "/": id });
     const charm = this.charms
       .get()
-      .find(({ cell }) => JSON.stringify(cell.entityId) === JSON.stringify({ "/": id }));
+      .find((charm) => JSON.stringify(getEntityId(charm)) === idAsDocId);
     if (!charm) return undefined;
 
     // Make sure we have the recipe so we can run it!
-    await this.syncRecipe(charm.cell);
+    await this.syncRecipe(charm);
 
     // Make sure the charm is running. This is re-entrant and has no effect if
     // the charm is already running.
-    return run(undefined, undefined, charm.cell);
+    return run(undefined, undefined, charm.getAsDocLink().cell).asCell([], undefined, charmSchema);
```

Comment:
I feel like we should separate run and get - as you might want to get it without running ?  or does that complicate things

Reply:
I'm mostly thinking about the case of showing related charms / grouping all your previous charms together - if we end up having `parent` as an attribute - we would want to walk the parents but not actually run them

Reply:
yes, let's do that in another PR?